### ISSUE TITLE
Changing member password in the backoffice not working as expected when allowManuallyChangingPassword is set to true

### ISSUE THREAD

**Issue by HalldorLyngmo** [umbraco/Umbraco-CMS#4144]
It used to be possible to set the allowManuallyChangingPassword setting to true and be able to change a member password without needing their old password in the backoffice.

In Umbraco v7.13.0 this won't work and there is an error message complaining about the old password for the user even if the field is not there.

![image](https://user-images.githubusercontent.com/42831978/51393868-0efc1c80-1b39-11e9-9ba4-b48d9a27983e.png)

Reproduction
------------

1. Download Umbraco 7.13.0 or 7.13.1
2. Start up the Umbraco CMS and go ahead and create a member. 
The member property tab will then look like this:
![image](https://user-images.githubusercontent.com/42831978/51394028-8b8efb00-1b39-11e9-9b76-e0d8c9e45d11.png)

3. Go to the web.config and change allowManuallyChangingPassword="false" to allowManuallyChangingPassword="true"

Then the member properties tab will look like this:
![image](https://user-images.githubusercontent.com/42831978/51394085-c4c76b00-1b39-11e9-817b-de1b2e6d91c0.png)

4. Try to change the password then there will be an error message connected to validating the old password:
 ![image](https://user-images.githubusercontent.com/42831978/51393868-0efc1c80-1b39-11e9-9ba4-b48d9a27983e.png)

### Bug summary

It seems like the validation for changing the passwords is not working correctly. I think it might be connected to this issue that was fixed for 7.13.0

3069 - Current password field should be hidden when changing passwords and not applicable
 https://github.com/umbraco/Umbraco-CMS/issues/3069

The fix would be to change validation when changing the member password. If the allowManuallyChangingPassword is set to true it should not need the old password.

### Specifics

Umbraco v7.13.0 and v7.13.1
Using latest version of Chrome , did not check in other browsers

### Expected result

I would expect to be able to change the password for that member without needing the old password if this setting is true.

allowManuallyChangingPassword="true"

### Actual result

Unable to change the password from the backoffice without the old password

**Comment by jfsaliba** [comment]
Adding my voice here as well. Can't reset passwords without providing the "old" password ever since we upgraded to 7.13.0. Can't set a new password from the backoffice like described here, nor can we change a user's password in code without knowing the old password. The "reset password" option in the backoffice to generate a new password *does* still work however.

This is a huge problem since we allow users to reset their passwords themselves with emailed password reset links, and used to use the MembershipHelper's ChangePassword() method (just providing a ChangingPasswordModel object with only a NewPassword specified) to set the new password.

My workaround right now is to use the Membership Provider to do a ResetPassword() on the member and use that generated password for the OldPassword property in the ChangingPasswordModel, and then passing that all along to the Membership Provider's ChangePassword() method as usual. Works, but man, that is a rude surprise for a Friday morning! :grimacing:

**Comment by MitchOkontaSpire** [comment]
Same issue here. v7.12.4 works.

**Comment by skttl** [comment]
The same thing happens for users. I can't change my own password from neither the dialog from the avatar, nor in the users section.

I can change other users passwords though.

**Comment by skttl** [comment]
Maybe this one https://github.com/umbraco/Umbraco-CMS/blob/0150b972780cfa39cce58003a92f0f5b9909fb10/src/Umbraco.Web/Editors/PasswordChanger.cs#L43 is missing a check for `allowManuallyChangingPasswords=true`?

It just returns a failed attempt if the old password is null, which it is going to be if you don't show the old password field.

**Comment by adamhearn** [comment]
I also looked at the codebase yesterday and @skttl is correct, the problem lies in the recent changes to ChangePasswordWithMembershipProvider.

There used to be a block of code commented as "//This is an edge case and is only necessary for backwards compatibility:" but this was removed when the method was refactored. A similar code block exists at line 186 but this is "protected" by the reset property.

I'd fix it if I knew the intent but it's not clear how it should work. It may be that the angular directive on the password fields should not show/hide based on the checkbox value change or it could be that the method should indeed support the "edge case" as it did?

A workaround for a member administrator, albeit not a great one, is to enter the new password in the password field (no need to confirm it) then check the reset password checkbox. The angular directive will hide the password/confirm password fields but their values remain in the model. The revised logic [lines 180-210] will set the password to the value presented.

Hope that helps.

**Comment by Newnab** [comment]
Can confirm this is also an issue for us and our clients, is there a fix in the works?

**Comment by karaokid** [comment]
Same here, also an issue for our clients so we are interested in a solution, thank you.

**Comment by ronaldbarendse** [comment]
Would be nice if this gets property fixed!

For now, setting `allowManuallyChangingPassword` back to `false`, resetting the password and then use that generated password as 'Old password' seems the easiest word-around...

This work-around only works if `enablePasswordReset` is left enabled and also shows how flawed this logic is: if you can reset a password, it's also possible to manually reset it, right? 😕

**Comment by JohnBergman** [comment]
Just wondered if this make it into 7.15.0(or 7.14.1 if you decide to do a minor)?  We have a client facing mobile app that depends on this functionality for self service, I currently don't see it in the list

**Comment by sumedhashyam** [comment]
Any solution on this thread. Our clients are also facing the same issue.

**Comment by warrenbuckley** [comment]
Been playing spot the difference this morning - will investigate further when the regression happened to do some more detective work 🕵 🔍 

![image](https://user-images.githubusercontent.com/1389894/57364857-db8edd80-717b-11e9-8942-59f6f20a6a3a.png)

**Comment by warrenbuckley** [comment]
Found the commit that removed this b0292159331f62ca755fa855b5ed30a98974387d
@Shazwazza I have added this back in the PR #5423 for this, as I assume this was simply removed by mistake?

**Comment by Shazwazza** [comment]
@warrenbuckley I think it's because the password should be changed on this line, not where the code that you've put back in executes https://github.com/umbraco/Umbraco-CMS/pull/5423/files#diff-8529397cc05f7165e5300ea16e775622R193

The code you put back in was marked that it was there for legacy reasons, I vaguely remember this stuff, but will run some tests to see whats going on.

**Comment by Shazwazza** [comment]
## User membership provider

These are the default settings that we ship:

```xml
<add name="UsersMembershipProvider" type="Umbraco.Web.Security.Providers.UsersMembershipProvider, Umbraco" />
```

It is highly recommended to leave this as the default setting, we are transitioning away from membership providers which has been done for Users for many versions now. However due to all sorts of backwards compatibility scenarios, we still have to have this hybrid system.

When `allowManuallyChangingPassword` or `useLegacyEncoding` is true for this provider, then it will fallback to using legacy logic. This is due to backwards compatibility for upgrades and if you are upgrading old systems then you may need to leave these legacy settings in there.

User passwords are governed by ASP.Net Identity except in the case where we need to fallback to legacy logic (with either of those 2 settings), in which case password logic is governed by Membership providers. 

### Expected Behavior

#### Legacy settings

_This is when you have either `allowManuallyChangingPassword` or `useLegacyEncoding` set to true_

* As a user that has access to the user section, you can change other users' passwords without supplying an existing password
* When changing your own password you don't need to supply your existing password when allowManuallyChangingPassword is true ... this is a security issue which is why this is legacy and should not be used for the user provider.

✔️ When allowManuallyChangingPassword is true you can change other user passwords
❌ When allowManuallyChangingPassword is true you cannot change your own password, this is a bug 

#### Default settings

* As a user that has access to the user section, you can change other users' passwords
* When changing your own password, you will need to provide your current password

✔️ I have tested and confirmed this is working

#### Other notes

@jfsaliba regarding what you say about "nor can we change a user's password in code without knowing the old password." ... this is true, no Microsoft APIs (either ASP.Net Identity nor Membership provider) let you change a password without using the current one. So yes the trick is as you suggest if you must be able to forcefully do this. The same trick is used for both ASP.Net Identity and Membership providers APIs. Due to the nature of having to support both frameworks and backwards compatibility for so many scenarios, we don't have a public API that people can just use to do this right now, but perhaps we should look at cleaning up and making public the `Umbraco.Web.Editors.PasswordChanger` - it needs some work to make public though.

## Members membership provider

These are the default settings that we ship:

```xml
<add name="UmbracoMembershipProvider" type="Umbraco.Web.Security.Providers.MembersMembershipProvider, Umbraco" 
minRequiredNonalphanumericCharacters="0" minRequiredPasswordLength="10" useLegacyEncoding="false" 
enablePasswordRetrieval="false" enablePasswordReset="false" requiresQuestionAndAnswer="false" defaultMemberTypeAlias="Member" 
passwordFormat="Hashed" allowManuallyChangingPassword="false" />
```

Members passwords are governed entirely by Membership providers (unless you install my UmbracoIdentity package... but that's a different story!) which means we have a bit less to worry about with backwards compat and overlapping scenarios with aspnet identity and membership providers. 

### Expected Behavior

#### Default settings

With the default settings above, to change a members password you would have to check the 'reset' password checkbox, save the member, copy the generated password that it shows you, then click change password again, paste the value into the 'old password' field and fill in the new password field.

✔️ I have tested and confirmed this is working

#### allowManuallyChangingPassword = true

When this is true along with the default settings above, you will see a checkbox to reset the password along with a new/config password field (no old password). You should be able to just change the password to what you want.

❌ I can confirm this is not working, it is a bug

## Fixing

Part of the reason for this regression is because I thought we had updated the angular view to pass through the correct information to the PasswordChanger for both Users and Members, but it turns out to only be the case in one scenario. Like @warrenbuckley mentions above (https://github.com/umbraco/Umbraco-CMS/issues/4144#issuecomment-490426965), if we put the removed code back in, the above problems go away. However, I'm not convinced that is the correct way to solve this issue.

The password should be changed without putting that code block back in if the angular app passes through the correct info, i believe that is the underlying issue and what I assumed was already in place before. 

* ✔️ I've fixed the members issue above in rev f952877d04b4f88486b83e8a9709c96426e8791e
* ✔️ I've fixed the user issue above in rev 35645672aeb960e92c5984d6b77b71e99dfc5b33

## Next steps

The default members settings above is not ideal, an admin shouldn't have to jump through hoops to change a member password or rely on a dev to set the allowManuallyChangingPassword = true for the membership provider. 

We should update the logic to be like users where any admin that has access to the edit members can just change their passwords so we should ship the membership provider with allowManuallyChangingPassword=true by default.  Thoughts? 

Optionally... we can always hide the reset password checkbox for members (just like we do for users), since i would assume that admins wouldn't use this checkbox ever if they can just change a members password. Thoughts?

**Comment by jfsaliba** [comment]
@Shazwazza thanks for the detailed write-up! I was mincing terms earlier, I've only been referring to Members (not Users) so your section on Members membership providers is exactly right.

I agree for Members, if an admin user has permission to view and alter the member's information, they should be able to change/set their password straightaway. I always enabled allowManuallyChangingPassword by default on my installs (for admins, and for my described mechanism for members to self-reset their passwords).

**Comment by Newnab** [comment]
Regarding the following:



"Optionally... we can always hide the reset password checkbox for members (just like we do for users), since i would assume that admins wouldn't use this checkbox ever if they can just change a members password. Thoughts?"



My two cents: I feel like your assumption is wrong, and the reset checkbox is a better (and certainly more natural to me) action for admins to take - triggering a reset and allowing members to set their own private passwords feels like better security practice to me, compared to an admin setting it to something, having to convey the password via some other means and then expecting a member to change it later to ensure it's private. Setting a member's password for them and emailing it to them separately always feels wrong to me.





---- On Thu, 09 May 2019 17:10:44 +0100 Shannon Deminick <notifications@github.com> wrote ----



User membership provider

These are the default settings that we ship:

<add name="UsersMembershipProvider" type="Umbraco.Web.Security.Providers.UsersMembershipProvider, Umbraco" />


It is highly recommended to leave this as the default setting, we are transitioning away from membership providers which has been done for Users for many versions now. However due to all sorts of backwards compatibility scenarios, we still have to have this hybrid system.

When allowManuallyChangingPassword or useLegacyEncoding is true for this provider, then it will fallback to using legacy logic. This is due to backwards compatibility for upgrades and if you are upgrading old systems then you may need to leave these legacy settings in there.

User passwords are governed by ASP.Net Identity except in the case where we need to fallback to legacy logic (with either of those 2 settings), in which case password logic is governed by Membership providers.

Expected Behavior

Legacy settings

This is when you have either allowManuallyChangingPassword or useLegacyEncoding set to true

As a user that has access to the user section, you can change other users' passwords without supplying an existing password

When changing your own password you don't need to supply your existing password when allowManuallyChangingPassword is true ... this is a security issue which is why this is legacy and should not be used for the user provider.


✔️ When allowManuallyChangingPassword is true you can change other user passwords
 ❌ When allowManuallyChangingPassword is true you cannot change your own password, this is a bug
Default settings

As a user that has access to the user section, you can change other users' passwords

When changing your own password, you will need to provide your current password


✔️ I have tested and confirmed this is working

Other notes

https://github.com/jfsaliba regarding what you say about "nor can we change a user's password in code without knowing the old password." ... this is true, no Microsoft APIs (either ASP.Net Identity nor Membership provider) let you change a password without using the current one. So yes the trick is as you suggest if you must be able to forcefully do this. The same trick is used for both ASP.Net Identity and Membership providers APIs. Due to the nature of having to support both frameworks and backwards compatibility for so many scenarios, we don't have a public API that people can just use to do this right now, but perhaps we should look at cleaning up and making public the Umbraco.Web.Editors.PasswordChanger - it needs some work to make public though.

Members membership provider

These are the default settings that we ship:

<add name="UmbracoMembershipProvider" type="Umbraco.Web.Security.Providers.MembersMembershipProvider, Umbraco" 

minRequiredNonalphanumericCharacters="0" minRequiredPasswordLength="10" useLegacyEncoding="false" 

enablePasswordRetrieval="false" enablePasswordReset="false" requiresQuestionAndAnswer="false" defaultMemberTypeAlias="Member" 

passwordFormat="Hashed" allowManuallyChangingPassword="false" />


Members passwords are governed entirely by Membership providers (unless you install my UmbracoIdentity package... but that's a different story!) which means we have a bit less to worry about with backwards compat and overlapping scenarios with aspnet identity and membership providers.

Expected Behavior

Default settings

With the default settings above, to change a members password you would have to check the 'reset' password checkbox, save the member, copy the generated password that it shows you, then click change password again, paste the value into the 'old password' field and fill in the new password field.

✔️ I have tested and confirmed this is working

allowManuallyChangingPassword = true

When this is true along with the default settings above, you will see a checkbox to reset the password along with a new/config password field (no old password). You should be able to just change the password to what you want.

❌ I can confirm this is not working, it is a bug

Fixing

Part of the reason for this regression is because I thought we had updated the angular view to pass through the correct information to the PasswordChanger for both Users and Members, but it turns out to only be the case in one scenario. Like https://github.com/warrenbuckley mentions above (https://github.com/umbraco/Umbraco-CMS/issues/4144#issuecomment-490426965), if we put the removed code back in, the above problems go away. However, I'm not convinced that is the correct way to solve this issue.

The password should be changed without putting that code block back in if the angular app passes through the correct info, i believe that is the underlying issue and what I assumed was already in place before.

✔️ I've fixed the members issue above in rev https://github.com/umbraco/Umbraco-CMS/commit/f952877d04b4f88486b83e8a9709c96426e8791e

✔️ I've fixed the user issue above in rev https://github.com/umbraco/Umbraco-CMS/commit/35645672aeb960e92c5984d6b77b71e99dfc5b33


Next steps

The default members settings above is not ideal, an admin shouldn't have to jump through hoops to change a member password or rely on a dev to set the allowManuallyChangingPassword = true for the membership provider.

We should update the logic to be like users where any admin that has access to the edit members can just change their passwords so we should ship the membership provider with allowManuallyChangingPassword=true by default.  Thoughts?

Optionally... we can always hide the reset password checkbox for members (just like we do for users), since i would assume that admins wouldn't use this checkbox ever if they can just change a members password. Thoughts?

**Comment by Shazwazza** [comment]
@Newnab thanks for the input! I knew assuming was going to bite me back ;) I agree with you since that is how i personally would want to work but I just assumed that admins might not. It's easier to leave the reset checkbox anyways and provide that option so I won't make any changes there.

@jfsaliba thanks for the feedback too, for now I'm going to leave the default member provider config as-is but I'll create a new issue can be opened up to set allowManuallyChangingPassword=true by default

So with that i just need @warrenbuckley to re-run some tests and in theory we can get these fixes merged in.

**Comment by warrenbuckley** [comment]
All fixed in PR #5423
